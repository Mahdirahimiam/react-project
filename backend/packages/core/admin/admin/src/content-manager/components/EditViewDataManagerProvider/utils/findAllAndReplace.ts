import type { FormattedComponentLayout } from '../../../utils/layouts';
import type { Attribute, Common } from '@strapi/types';

type ReplacementFn<
  TData extends Record<string, any>,
  TKey extends keyof TData,
  TReplacementValue
> = (data: TData[TKey], args: { path: string[]; parent: TData }) => TReplacementValue;

/**
 * This function will recursively find everything and replace it with a value
 * based on the boolean return of the predicate function e.g. `type === 'relation'`.
 *
 * If you provide a function it will call that function with data value you're replacing with
 * a second argument with the path to the value and it's parent.
 *
 * It's original use was for the preperation of action items for the INIT_FORM action. It requires
 * knowledge of the `components` in the entity, however `components` doesn't change nor does the predicate
 * function so we don't need to pass it everytime hence why it's curried.
 *
 */
const findAllAndReplaceSetup = <
  TReplacementValue,
  TData extends Record<string, any> = Record<string, any>
>(
  components: Record<string, FormattedComponentLayout>,
  predicate: (
    value: FormattedComponentLayout['attributes'][string],
    args: { path: string[]; parent: { [key: string]: Attribute.Any } | Attribute.Any }
  ) => boolean = () => false,
  replacement?: ReplacementFn<TData, keyof TData, TReplacementValue> | TReplacementValue
) => {
  const findAllAndReplace = <TData extends Record<string, any> = Record<string, any>>(
    data: TData,
    attributes: { [key: string]: Attribute.Any },
    {
      ignoreFalseyValues = false,
      path = [],
      parent = attributes,
    }: {
      ignoreFalseyValues?: boolean;
      path?: string[];
      parent?: { [key: string]: Attribute.Any } | Attribute.Any;
    } = {}
  ) => {
    return Object.entries(attributes).reduce<Record<string, any>>(
      (acc, [key, value]) => {
        if (
          ignoreFalseyValues &&
          (acc === null || acc === undefined || acc[key] === undefined || acc[key] === null)
        ) {
          return acc;
        }

        if (predicate(value, { path: [...path, key], parent })) {
          acc[key] =
            typeof replacement === 'function'
              ? // @ts-expect-error â€“ TODO: Fix this.
                replacement(acc[key], { path: [...path, key], parent: acc })
              : replacement;
        }

        if (value.type === 'component') {
          const componentAttributes = components[value.component].attributes;

          if (!value.repeatable && acc[key] && typeof acc[key] === 'object') {
            acc[key] = findAllAndReplace(acc[key], componentAttributes, {
              ignoreFalseyValues,
              path: [...path, key],
              parent: attributes[key],
            });
          } else if (value.repeatable && Array.isArray(acc[key])) {
            acc[key] = (
              acc[key] as Attribute.GetValue<Attribute.Component<Common.UID.Component, true>>
            ).map((datum, index) => {
              const data = findAllAndReplace(datum, componentAttributes, {
                ignoreFalseyValues,
                path: [...path, key, index.toString()],
                parent: attributes[key],
              });

              return data;
            });
          }
        } else if (value.type === 'dynamiczone' && Array.isArray(acc[key])) {
          acc[key] = (acc[key] as Attribute.GetValue<Attribute.DynamicZone>).map((datum, index) => {
            const componentAttributes = components[datum.__component].attributes;
            const data = findAllAndReplace(datum, componentAttributes, {
              ignoreFalseyValues,
              path: [...path, key, index.toString()],
              parent: attributes[key],
            });

            return data;
          });
        }

        return acc;
      },
      { ...data }
    );
  };

  return findAllAndReplace;
};

export { findAllAndReplaceSetup as findAllAndReplace };
