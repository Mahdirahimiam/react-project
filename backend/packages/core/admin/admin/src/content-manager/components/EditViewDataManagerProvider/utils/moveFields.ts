const moveFields = <TValue>(initialValue: TValue[], from: number, to: number, value: TValue) => {
  const returnedValue = initialValue.slice();

  returnedValue.splice(from, 1);
  returnedValue.splice(to, 0, value);

  return returnedValue;
};

export { moveFields };
