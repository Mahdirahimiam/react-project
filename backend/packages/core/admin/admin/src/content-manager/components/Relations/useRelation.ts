import { useEffect, useState } from 'react';

import { useCallbackRef, useFetchClient } from '@strapi/helper-plugin';
import { useInfiniteQuery } from 'react-query';

import {
  NormalizeRelationArgs,
  NormalizedRelation,
  normalizeRelations,
} from './utils/normalizeRelations';

import type { Contracts } from '@strapi/plugin-content-manager/_internal/shared';

interface UseRelationArgs {
  relation: {
    enabled: boolean;
    endpoint: string;
    normalizeArguments: NormalizeRelationArgs;
    onLoad: (data: NormalizedRelation[]) => void;
    pageParams?: Record<string, any>;
    pageGoal?: number;
  };
  search: {
    endpoint: string;
    pageParams?: Record<string, any>;
  };
}

const useRelation = (cacheKey: any[] = [], { relation, search }: UseRelationArgs) => {
  const [searchParams, setSearchParams] = useState({});
  const [currentPage, setCurrentPage] = useState(0);
  const { get } = useFetchClient();

  const { onLoad: onLoadRelations, normalizeArguments } = relation;

  const relationsRes = useInfiniteQuery(
    ['relation', ...cacheKey],
    async ({ pageParam = 1 }) => {
      try {
        const { data } = await get<Contracts.Relations.FindExisting.Response>(relation?.endpoint, {
          params: {
            ...(relation.pageParams ?? {}),
            page: pageParam,
          },
        });

        setCurrentPage(pageParam);

        return data;
      } catch (err) {
        return null;
      }
    },
    {
      cacheTime: 0,
      enabled: relation.enabled,
      getNextPageParam(lastPage) {
        const isXToOneRelation = lastPage && !('pagination' in lastPage);

        if (
          !lastPage || // the API may send an empty 204 response
          isXToOneRelation || // xToOne relations do not have a pagination
          lastPage?.pagination.page >= lastPage?.pagination.pageCount
        ) {
          return undefined;
        }

        // eslint-disable-next-line consistent-return
        return lastPage.pagination.page + 1;
      },
      select: (data) => ({
        ...data,
        pages: data.pages.map((page) => {
          if (!page) {
            return page;
          }

          let normalizedResults: Contracts.Relations.RelationResult[] = [];

          // xToOne relations return an object, which we normalize so that relations
          // always have the same shape
          if ('data' in page && page.data) {
            normalizedResults = [page.data];
          } else if ('results' in page && page.results) {
            normalizedResults = [...page.results].reverse();
          }

          return {
            pagination: 'pagination' in page ? page.pagination : undefined,
            results: normalizedResults,
          };
        }),
      }),
    }
  );

  const { pageGoal } = relation;

  const { status, data, fetchNextPage, hasNextPage } = relationsRes;

  useEffect(() => {
    /**
     * This ensures the infiniteQuery hook fetching has caught-up with the modifiedData
     * state i.e. in circumstances where you add 10 relations, the browserState knows this,
     * but the hook would think it could fetch more, when in reality, it can't.
     */
    if (pageGoal && pageGoal > currentPage && hasNextPage && status === 'success') {
      fetchNextPage({
        pageParam: currentPage + 1,
      });
    }
  }, [pageGoal, currentPage, fetchNextPage, hasNextPage, status]);

  const onLoadRelationsCallback = useCallbackRef(onLoadRelations);

  useEffect(() => {
    if (status === 'success' && data && data.pages?.at(-1)?.results && onLoadRelationsCallback) {
      // everytime we fetch, we normalize prior to adding to redux
      const normalizedResults = normalizeRelations(
        data.pages.at(-1)?.results ?? [],
        normalizeArguments
      );

      // this is loadRelation from EditViewDataManagerProvider
      onLoadRelationsCallback(normalizedResults);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [status, onLoadRelationsCallback, data]);

  const searchRes = useInfiniteQuery(
    ['relation', ...cacheKey, 'search', JSON.stringify(searchParams)],
    async ({ pageParam = 1 }) => {
      try {
        const { data } = await get<Contracts.Relations.FindAvailable.Response>(search.endpoint, {
          params: {
            ...(search.pageParams ?? {}),
            ...searchParams,
            page: pageParam,
          },
        });

        return data;
      } catch (err) {
        return null;
      }
    },
    {
      enabled: Object.keys(searchParams).length > 0,
      getNextPageParam(lastPage) {
        if (
          !lastPage?.pagination ||
          (lastPage.pagination && lastPage.pagination.page >= lastPage.pagination.pageCount)
        ) {
          return undefined;
        }

        // eslint-disable-next-line consistent-return
        return lastPage.pagination.page + 1;
      },
    }
  );

  const searchFor = (term: string, options: object = {}) => {
    setSearchParams({
      ...options,
      _q: term,
      _filter: '$containsi',
    });
  };

  return { relations: relationsRes, search: searchRes, searchFor };
};

export { useRelation };
export type { UseRelationArgs };
