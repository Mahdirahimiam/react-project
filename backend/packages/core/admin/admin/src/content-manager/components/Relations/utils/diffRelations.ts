import { RelationData } from '../../EditViewDataManagerProvider/reducer';

export const diffRelations = (
  browserStateRelations: RelationData[] = [],
  serverStateRelations: RelationData[] = []
): [connected: RelationData['id'][], disconnected: RelationData['id'][]] => {
  const connected = browserStateRelations.reduce<RelationData['id'][]>((acc, relation) => {
    if (!serverStateRelations.find((oldRelation) => oldRelation.id === relation.id)) {
      return [...acc, relation.id];
    }

    return acc;
  }, []);

  const disconnected = serverStateRelations.reduce<RelationData['id'][]>((acc, relation) => {
    if (!browserStateRelations.find((oldRelation) => oldRelation.id === relation.id)) {
      return [...acc, relation.id];
    }

    return acc;
  }, []);

  return [connected, disconnected];
};
