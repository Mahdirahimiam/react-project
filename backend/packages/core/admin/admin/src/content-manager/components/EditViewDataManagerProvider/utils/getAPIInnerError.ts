import { getTranslation } from '../../../utils/translations';

import type { ApiError, TranslationMessage } from '@strapi/helper-plugin';
import type { AxiosError } from 'axios';

const getAPIInnerError = (error: AxiosError<{ error: ApiError }>) => {
  if (
    error.response &&
    typeof error.response.data.error.details === 'object' &&
    error.response.data.error.details !== null &&
    'errors' in error.response.data.error.details &&
    Array.isArray(error.response.data.error.details.errors)
  ) {
    const errorPayload = error.response.data.error.details.errors;
    const validationErrors = errorPayload.reduce<Record<string, TranslationMessage>>((acc, err) => {
      acc[err.path.join('.')] = {
        id: getTranslation(`apiError.${err.message}`),
        defaultMessage: err.message,
        values: {
          field: err.path[err.path.length - 1],
        },
      };

      return acc;
    }, {});

    return validationErrors;
  }

  return {};
};

export { getAPIInnerError };
