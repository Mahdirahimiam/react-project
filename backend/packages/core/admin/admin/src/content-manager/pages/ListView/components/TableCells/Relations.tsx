import * as React from 'react';

import { Typography, Badge, Flex, Loader, useNotifyAT } from '@strapi/design-system';
import { Menu } from '@strapi/design-system/v2';
import { useFetchClient } from '@strapi/helper-plugin';
import { useIntl } from 'react-intl';
import { useQuery } from 'react-query';
import styled from 'styled-components';

import { getTranslation } from '../../../../utils/translations';

import { CellValue } from './CellValue';

import type { CellContentProps } from './CellContent';
import type { Contracts } from '@strapi/plugin-content-manager/_internal/shared';
import type { Entity } from '@strapi/types';

/* -------------------------------------------------------------------------------------------------
 * RelationSingle
 * -----------------------------------------------------------------------------------------------*/

interface RelationSingleProps extends Pick<CellContentProps, 'metadatas' | 'content'> {}

const RelationSingle = ({ metadatas, content }: RelationSingleProps) => {
  return (
    <TypographyMaxWidth textColor="neutral800" ellipsis>
      <CellValue
        // integer is default because that's what the id will be.
        type={metadatas.mainField?.type ?? 'integer'}
        value={metadatas.mainField?.name ? content[metadatas.mainField?.name] : content.id}
      />
    </TypographyMaxWidth>
  );
};

const TypographyMaxWidth = styled(Typography)`
  max-width: 500px;
`;

/* -------------------------------------------------------------------------------------------------
 * RelationMultiple
 * -----------------------------------------------------------------------------------------------*/

interface RelationMultipleProps extends Pick<CellContentProps, 'metadatas' | 'name' | 'content'> {
  entityId: Entity.ID;
  uid: string;
}

const RelationMultiple = ({ metadatas, name, entityId, content, uid }: RelationMultipleProps) => {
  const { formatMessage } = useIntl();
  const { notifyStatus } = useNotifyAT();
  const [isOpen, setIsOpen] = React.useState(false);

  const { get } = useFetchClient();

  const [fieldName] = name.split('.');

  const { data, isLoading } = useQuery(
    [uid, entityId, fieldName],
    async () => {
      const { data } = await get<Contracts.Relations.FindExisting.Response>(
        `/content-manager/relations/${uid}/${entityId}/${fieldName}`
      );

      if ('data' in data && data.data) {
        return {
          results: [data.data],
        };
      }

      if ('results' in data) {
        return { results: data.results, pagination: data.pagination };
      }

      throw new Error(
        `/content-manager/relations/${uid}/${entityId}/${fieldName} returned an error object with a success code.`
      );
    },
    {
      enabled: isOpen,
      staleTime: 0,
      select: (data) => ({
        ...data,
        results: [...data.results].reverse(),
      }),
    }
  );

  React.useEffect(() => {
    if (data) {
      notifyStatus(
        formatMessage({
          id: getTranslation('DynamicTable.relation-loaded'),
          defaultMessage: 'Relations have been loaded',
        })
      );
    }
  }, [data, formatMessage, notifyStatus]);

  return (
    <Menu.Root onOpenChange={(isOpen) => setIsOpen(isOpen)}>
      <MenuTrigger onClick={(e) => e.stopPropagation()}>
        <Flex gap={1} wrap="nowrap">
          <Badge>{content.count}</Badge>
          {formatMessage(
            {
              id: 'content-manager.containers.ListPage.items',
              defaultMessage: '{number, plural, =0 {items} one {item} other {items}}',
            },
            { number: content.count }
          )}
        </Flex>
      </MenuTrigger>
      <Menu.Content>
        {isLoading && (
          <Menu.Item disabled>
            <Loader small>
              {formatMessage({
                id: getTranslation('ListViewTable.relation-loading'),
                defaultMessage: 'Relations are loading',
              })}
            </Loader>
          </Menu.Item>
        )}

        {data?.results && (
          <>
            {data.results.map((entry) => (
              <Menu.Item key={entry.id} disabled>
                <TypographyMaxWidth ellipsis>
                  <CellValue
                    type={metadatas.mainField?.type ?? 'integer'}
                    // @ts-expect-error – can't use a string to index the RelationResult object.
                    value={metadatas.mainField?.name ? entry[metadatas.mainField.name] : entry.id}
                  />
                </TypographyMaxWidth>
              </Menu.Item>
            ))}

            {data?.pagination && data?.pagination.total > 10 && (
              <Menu.Item
                aria-disabled
                aria-label={formatMessage({
                  id: getTranslation('ListViewTable.relation-more'),
                  defaultMessage: 'This relation contains more entities than displayed',
                })}
              >
                <Typography>…</Typography>
              </Menu.Item>
            )}
          </>
        )}
      </Menu.Content>
    </Menu.Root>
  );
};

/**
 * TODO: this needs to be solved in the Design-System
 */
const MenuTrigger = styled(Menu.Trigger)`
  svg {
    width: ${6 / 16}rem;
    height: ${4 / 16}rem;
  }
`;

export { RelationSingle, RelationMultiple };
export type { RelationSingleProps, RelationMultipleProps };
