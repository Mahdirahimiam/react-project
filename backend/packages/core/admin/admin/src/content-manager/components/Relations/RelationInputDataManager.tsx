import * as React from 'react';

import { NotAllowedInput, useCMEditViewDataManager } from '@strapi/helper-plugin';
import get from 'lodash/get';
import pick from 'lodash/pick';
import { MessageDescriptor, useIntl } from 'react-intl';
import { useRouteMatch } from 'react-router-dom';

import { getInitialDataPathUsingTempKeys } from '../../utils/paths';
import { getTranslation } from '../../utils/translations';

import { RelationInput, RelationInputProps } from './RelationInput';
import { useRelation } from './useRelation';
import { diffRelations } from './utils/diffRelations';
import { NormalizedRelation, normalizeRelation } from './utils/normalizeRelations';
import { normalizeSearchResults } from './utils/normalizeSearchResults';

import type { Contracts } from '@strapi/plugin-content-manager/_internal/shared';

const PUBLICATION_STATES = {
  DRAFT: 'draft',
  PUBLISHED: 'published',
} as const;

const RELATIONS_TO_DISPLAY = 5;

const SEARCH_RESULTS_TO_DISPLAY = 10;

interface RelationInputDataManagerProps
  extends Pick<
    RelationInputProps,
    'description' | 'error' | 'labelAction' | 'name' | 'required' | 'size'
  > {
  componentUid?: string;
  editable?: boolean;
  intlLabel: MessageDescriptor;
  isUserAllowedToEditField?: boolean;
  isUserAllowedToReadField?: boolean;
  mainField: {
    name: string;
  };
  placeholder?: MessageDescriptor;
  queryInfos: {
    defaultParams?: object;
    shouldDisplayRelationLink?: boolean;
  };
  relationType: string;
  targetModel: string;
}

const RelationInputDataManager = ({
  componentUid,
  name,
  error,
  editable,
  description,
  intlLabel,
  isUserAllowedToEditField,
  isUserAllowedToReadField,
  labelAction,
  mainField,
  placeholder,
  queryInfos: { defaultParams, shouldDisplayRelationLink = false },
  required,
  relationType,
  size,
  targetModel,
}: RelationInputDataManagerProps) => {
  const {
    isCreatingEntry,
    createActionAllowedFields,
    readActionAllowedFields,
    updateActionAllowedFields,
    slug,
    modifiedData,
    initialData,
    relationConnect,
    relationDisconnect,
    relationLoad,
    relationReorder,
  } = useCMEditViewDataManager();

  /**
   * This is our cloning route because the EditView & CloneView share the same UI component
   * We need the origin ID to pre-load the relations into the modifiedData of the new
   * to-be-cloned entity.
   */
  const { params } =
    useRouteMatch<{ origin?: string }>(
      '/content-manager/collection-types/:collectionType/create/clone/:origin'
    ) ?? {};

  const { origin } = params ?? {};
  const isCloningEntry = Boolean(origin);
  const isComponentRelation = Boolean(componentUid);

  const isFieldAllowed = React.useMemo(() => {
    if (isUserAllowedToEditField === true) {
      return true;
    }

    const allowedFields = isCreatingEntry ? createActionAllowedFields : updateActionAllowedFields;

    return allowedFields.includes(name);
  }, [
    isCreatingEntry,
    createActionAllowedFields,
    name,
    isUserAllowedToEditField,
    updateActionAllowedFields,
  ]);

  const isFieldReadable = React.useMemo(() => {
    if (isUserAllowedToReadField) {
      return true;
    }

    const allowedFields = isCreatingEntry ? [] : readActionAllowedFields;

    return allowedFields.includes(name);
  }, [isCreatingEntry, isUserAllowedToReadField, name, readActionAllowedFields]);

  const fieldNameKeys = name.split('.');
  const componentId = componentUid ? get(modifiedData, fieldNameKeys.slice(0, -1))?.id : undefined;

  const entityId = origin || modifiedData.id;

  // /content-manager/relations/[model]/[id]/[field-name]
  const relationFetchEndpoint = React.useMemo(() => {
    if (isCreatingEntry && !origin) {
      return null;
    }

    if (componentUid) {
      // repeatable components and dz are dynamically created
      // if no componentId exists in modifiedData it means that the user just created it
      // there then are no relations to request
      return componentId
        ? `/content-manager/relations/${componentUid}/${componentId}/${fieldNameKeys.at(-1)}`
        : null;
    }

    return `/content-manager/relations/${slug}/${entityId}/${name.split('.').at(-1)}`;
  }, [isCreatingEntry, origin, componentUid, slug, entityId, name, componentId, fieldNameKeys]);

  // /content-manager/relations/[model]/[field-name]
  const relationSearchEndpoint = React.useMemo(() => {
    if (componentUid) {
      return `/content-manager/relations/${componentUid}/${name.split('.').at(-1)}`;
    }

    return `/content-manager/relations/${slug}/${name.split('.').at(-1)}`;
  }, [componentUid, slug, name]);

  const [liveText, setLiveText] = React.useState('');
  const { formatMessage } = useIntl();

  const nameSplit = name.split('.');

  const initialDataPath = getInitialDataPathUsingTempKeys(initialData, modifiedData)(name);

  const relationsFromModifiedData = get(modifiedData, name, []);

  const currentLastPage = Math.ceil(get(initialData, name, []).length / RELATIONS_TO_DISPLAY);

  const { relations, search, searchFor } = useRelation(
    [slug, initialDataPath.join('.'), modifiedData.id, defaultParams],
    {
      relation: {
        enabled: !!relationFetchEndpoint,
        endpoint: relationFetchEndpoint!,
        pageGoal: currentLastPage,
        pageParams: {
          ...defaultParams,
          pageSize: RELATIONS_TO_DISPLAY,
        },
        onLoad(value) {
          relationLoad?.({
            target: {
              initialDataPath: ['initialData', ...initialDataPath],
              modifiedDataPath: ['modifiedData', ...nameSplit],
              value,
            },
          });
        },
        normalizeArguments: {
          mainFieldName: mainField.name,
          shouldAddLink: shouldDisplayRelationLink,
          targetModel,
        },
      },
      search: {
        endpoint: relationSearchEndpoint,
        pageParams: {
          ...defaultParams,
          // eslint-disable-next-line no-nested-ternary
          entityId:
            isCreatingEntry || isCloningEntry
              ? undefined
              : isComponentRelation
              ? componentId
              : entityId,
          pageSize: SEARCH_RESULTS_TO_DISPLAY,
        },
      },
    }
  );

  const isMorph = relationType.toLowerCase().includes('morph');
  const toOneRelation = [
    'oneWay',
    'oneToOne',
    'manyToOne',
    'oneToManyMorph',
    'oneToOneMorph',
  ].includes(relationType);

  const isDisabled = React.useMemo(() => {
    if (isMorph) {
      return true;
    }

    if (!isCreatingEntry) {
      return (!isFieldAllowed && isFieldReadable) || !editable;
    }

    return !editable;
  }, [isMorph, isCreatingEntry, editable, isFieldAllowed, isFieldReadable]);

  const handleRelationConnect = (relation: Contracts.Relations.RelationResult) => {
    /**
     * Any relation being added to the store should be normalized so it has it's link.
     */
    const normalizedRelation = normalizeRelation(relation, {
      mainFieldName: mainField.name,
      shouldAddLink: shouldDisplayRelationLink,
      targetModel,
    });

    relationConnect?.({ name, value: normalizedRelation, toOneRelation });
  };

  const handleRelationDisconnect = (relation: NormalizedRelation) => {
    relationDisconnect?.({ name, id: relation.id });
  };

  const handleRelationLoadMore = () => {
    relations.fetchNextPage();
  };

  const handleSearch = (term = '') => {
    const [connected, disconnected] = diffRelations(
      relationsFromModifiedData,
      get(initialData, name)
    );

    searchFor(term, {
      idsToInclude: disconnected,
      idsToOmit: connected,
    });
  };

  const handleSearchMore = () => {
    search.fetchNextPage();
  };

  const getItemPos = (index: number) => `${index + 1} of ${relationsFromModifiedData.length}`;

  const handleRelationReorder = (oldIndex: number, newIndex: number) => {
    const item = relationsFromModifiedData[oldIndex];

    setLiveText(
      formatMessage(
        {
          id: getTranslation('dnd.reorder'),
          defaultMessage: '{item}, moved. New position in list: {position}.',
        },
        {
          item: item.mainField ?? item.id,
          position: getItemPos(newIndex),
        }
      )
    );

    relationReorder?.({
      name,
      newIndex,
      oldIndex,
    });
  };

  const handleGrabItem = (index: number) => {
    const item = relationsFromModifiedData[index];

    setLiveText(
      formatMessage(
        {
          id: getTranslation('dnd.grab-item'),
          defaultMessage: `{item}, grabbed. Current position in list: {position}. Press up and down arrow to change position, Spacebar to drop, Escape to cancel.`,
        },
        {
          item: item.mainField ?? item.id,
          position: getItemPos(index),
        }
      )
    );
  };

  const handleDropItem = (index: number) => {
    const item = relationsFromModifiedData[index];

    setLiveText(
      formatMessage(
        {
          id: getTranslation('dnd.drop-item'),
          defaultMessage: `{item}, dropped. Final position in list: {position}.`,
        },
        {
          item: item.mainField ?? item.id,
          position: getItemPos(index),
        }
      )
    );
  };

  const handleCancel = (index: number) => {
    const item = relationsFromModifiedData[index];

    setLiveText(
      formatMessage(
        {
          id: getTranslation('dnd.cancel-item'),
          defaultMessage: '{item}, dropped. Re-order cancelled.',
        },
        {
          item: item.mainField ?? item.id,
        }
      )
    );
  };

  if (
    (!isFieldAllowed && isCreatingEntry) ||
    (!isCreatingEntry && !isFieldAllowed && !isFieldReadable)
  ) {
    return <NotAllowedInput name={name} intlLabel={intlLabel} labelAction={labelAction} />;
  }

  /**
   * How to calculate the total number of relations even if you don't
   * have them all loaded in the browser.
   *
   * 1. The `infiniteQuery` gives you the total number of relations in the pagination result.
   * 2. You can diff the length of the browserState vs the fetchedServerState to determine if you've
   * either added or removed relations.
   * 3. Add them together, if you've removed relations you'll get a negative number and it'll
   * actually subtract from the total number on the server (regardless of how many you fetched).
   */
  const browserRelationsCount = relationsFromModifiedData.length;
  const serverRelationsCount = (get(initialData, initialDataPath) ?? []).length;
  const realServerRelationsCount = relations.data?.pages[0]?.pagination?.total ?? 0;
  /**
   * _IF_ theres no relations data and the browserCount is the same as serverCount you can therefore assume
   * that the browser count is correct because we've just _made_ this entry and the in-component hook is now fetching.
   */
  const totalRelations =
    !relations.data && browserRelationsCount === serverRelationsCount
      ? browserRelationsCount
      : browserRelationsCount - serverRelationsCount + realServerRelationsCount;

  return (
    <RelationInput
      error={error}
      canReorder={!toOneRelation}
      description={description}
      disabled={isDisabled}
      iconButtonAriaLabel={formatMessage({
        id: getTranslation('components.RelationInput.icon-button-aria-label'),
        defaultMessage: 'Drag',
      })}
      id={name}
      label={`${formatMessage({
        id: intlLabel.id,
        defaultMessage: intlLabel.defaultMessage,
      })} ${totalRelations > 0 ? `(${totalRelations})` : ''}`}
      labelAction={labelAction}
      labelLoadMore={
        !isCreatingEntry || isCloningEntry
          ? formatMessage({
              id: getTranslation('relation.loadMore'),
              defaultMessage: 'Load More',
            })
          : undefined
      }
      labelDisconnectRelation={formatMessage({
        id: getTranslation('relation.disconnect'),
        defaultMessage: 'Remove',
      })}
      listAriaDescription={formatMessage({
        id: getTranslation('dnd.instructions'),
        defaultMessage: `Press spacebar to grab and re-order`,
      })}
      liveText={liveText}
      loadingMessage={formatMessage({
        id: getTranslation('relation.isLoading'),
        defaultMessage: 'Relations are loading',
      })}
      name={name}
      noRelationsMessage={formatMessage({
        id: getTranslation('relation.notAvailable'),
        defaultMessage: 'No relations available',
      })}
      numberOfRelationsToDisplay={RELATIONS_TO_DISPLAY}
      onDropItem={handleDropItem}
      onGrabItem={handleGrabItem}
      onCancel={handleCancel}
      onRelationConnect={handleRelationConnect}
      onRelationDisconnect={handleRelationDisconnect}
      onRelationLoadMore={handleRelationLoadMore}
      onRelationReorder={handleRelationReorder}
      onSearch={(term) => handleSearch(term)}
      onSearchNextPage={() => handleSearchMore()}
      placeholder={formatMessage(
        placeholder || {
          id: getTranslation('relation.add'),
          defaultMessage: 'Add relation',
        }
      )}
      publicationStateTranslations={{
        [PUBLICATION_STATES.DRAFT]: formatMessage({
          id: getTranslation('relation.publicationState.draft'),
          defaultMessage: 'Draft',
        }),

        [PUBLICATION_STATES.PUBLISHED]: formatMessage({
          id: getTranslation('relation.publicationState.published'),
          defaultMessage: 'Published',
        }),
      }}
      relations={pick(
        { ...relations, data: relationsFromModifiedData },
        'data',
        'hasNextPage',
        'isFetchingNextPage',
        'isLoading',
        'isSuccess'
      )}
      required={required}
      searchResults={normalizeSearchResults(search, {
        mainFieldName: mainField.name,
      })}
      size={size}
    />
  );
};

export { RelationInputDataManager, PUBLICATION_STATES };
export type { RelationInputDataManagerProps };
