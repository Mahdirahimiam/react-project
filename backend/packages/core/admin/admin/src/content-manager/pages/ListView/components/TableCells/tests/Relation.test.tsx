import { render } from '@tests/utils';

import { RelationSingle, RelationMultiple } from '../Relations';

describe('Relation cell content', () => {
  describe('RelationSingle', () => {
    const DEFAULT_PROPS_FIXTURE = {
      metadatas: {
        label: '',
        mainField: {
          name: 'name',
          type: 'string' as const,
        },
      },
      content: {
        name: 1,
      },
    };

    it('renders and matches the snapshot', async () => {
      const { container } = render(<RelationSingle {...DEFAULT_PROPS_FIXTURE} />);
      expect(container).toMatchSnapshot();
    });
  });
  describe('RelationMultiple', () => {
    const DEFAULT_PROPS_FIXTURE = {
      uid: 'api::address.address' as const,
      entityId: 1,
      metadatas: {
        label: '',
        mainField: {
          name: 'name',
          type: 'string' as const,
        },
      },
      content: {
        count: 1,
      },
      name: 'categories.name',
      rowId: 1,
    };
    it('renders and renders the menu when clicked', async () => {
      const { getByRole, findByRole, user } = render(
        <RelationMultiple {...DEFAULT_PROPS_FIXTURE} />
      );

      expect(getByRole('button', { name: '1 item' })).toBeInTheDocument();

      await user.click(getByRole('button', { name: '1 item' }));

      expect(getByRole('menu')).toBeInTheDocument();

      await findByRole('menuitem', { name: 'Relation entity 1' });

      [1, 2, 3].forEach((number) => {
        expect(getByRole('menuitem', { name: `Relation entity ${number}` })).toBeInTheDocument();
      });
    });

    it('Displays related entities in reversed order', async () => {
      const { user, getByRole, getAllByRole } = render(
        <RelationMultiple {...DEFAULT_PROPS_FIXTURE} />
      );

      await user.click(getByRole('button', { name: '1 item' }));

      getAllByRole('menuitem').forEach((button, i) => {
        expect(button).toHaveTextContent(`Relation entity ${3 - i}`);
      });
    });
  });
});
