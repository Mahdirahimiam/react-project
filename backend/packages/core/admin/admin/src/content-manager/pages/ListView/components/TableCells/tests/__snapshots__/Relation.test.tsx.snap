// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Relation cell content RelationSingle renders and matches the snapshot 1`] = `
.c4 {
  border: 0;
  -webkit-clip: rect(0 0 0 0);
  clip: rect(0 0 0 0);
  height: 1px;
  margin: -1px;
  overflow: hidden;
  padding: 0;
  position: absolute;
  width: 1px;
}

.c2 {
  font-size: 0.875rem;
  line-height: 1.43;
  display: block;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  color: #32324d;
}

.c0 {
  margin-left: -250px;
  position: fixed;
  left: 50%;
  top: 2.875rem;
  z-index: 10;
  width: 31.25rem;
}

.c1 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 8px;
}

.c3 {
  max-width: 500px;
}

<div>
  <div
    class="c0 c1"
  />
  <span
    class="c2 c3"
  >
    1
  </span>
  <div
    class="c4"
  >
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-log"
      role="log"
    />
    <p
      aria-live="polite"
      aria-relevant="all"
      id="live-region-status"
      role="status"
    />
    <p
      aria-live="assertive"
      aria-relevant="all"
      id="live-region-alert"
      role="alert"
    />
  </div>
</div>
`;
